mangaki/mangaki/utils/anidb.py

```diff
@@ -160,7 +160,95 @@ def get_xml(self, anidb_aid: int):
+            title = str(title_node.string).strip()
+            lang = title_node.get('xml:lang')
+            title_type = title_node.get('type')
+
+            titles.append({
+                'title': title,
+                'lang': lang,
+                'type': title_type
+            })
+
+            if title_type == 'main':
+                main_title = title
+
+        return titles, main_title
+
+    def get_creators(self, anidb_aid=None, creators_soup=None):
+        if anidb_aid is not None:
+            anime = self.get_xml(anidb_aid)
+            creators_soup = anime.creators
+
+        creators = []
+        studio = None
+        # FIXME: cache this query
+        staff_map = dict(Role.objects.values_list('slug', 'pk'))
+        for creator_node in creators_soup.find_all('name'):
+            creator = str(creator_node.string).strip()
+            creator_id = creator_node.get('id')
+            creator_type = creator_node.get('type')
+            staff_id = None
+
+            if creator_type == 'Direction':
+                staff_id = 'director'
+            elif creator_type == 'Music':
+                staff_id = 'composer'
+            elif creator_type == 'Original Work' or creator_type == 'Story Composition':
+                staff_id = 'author'
+            elif creator_type == 'Animation Work' or creator_type == 'Work':
+                # AniDB marks Studio as such a creator's type
+                studio = Studio.objects.filter(title=creator).first()
+                if studio is None:
+                    studio = Studio(title=creator)
+                    studio.save()
+
+            if staff_id is not None:
+                creators.append({
+                    "role_id": staff_map[staff_id],
+                    "name": creator,
+                    "anidb_creator_id": creator_id
+                })
+
+        return creators, studio
+
+    def _build_staff(self,
+                     work: Work,
+                     creators: List[Dict[str, Any]]) -> List[Staff]:
+        artists_to_add = []
+        artists_list = []
+        for nc in creators:
+            artist = Artist.objects.filter(Q(name=nc["name"]) | Q(anidb_creator_id=nc["anidb_creator_id"])).first()
+
+            if not artist: # This artist does not yet exist : will be bulk created
+                artist = Artist(name=nc["name"], anidb_creator_id=nc["anidb_creator_id"])
+                artists_to_add.append(artist)
+            else: # This artist exists : prevent duplicates by updating with the AniDB id
+                artist.name = nc["name"]
+                artist.anidb_creator_id = nc["anidb_creator_id"]
+                artist.save()
+            artists_list.append(artist)
+
+        Artist.objects.bulk_create(artists_to_add)
+
+        staffs_to_add = []
+        for index, nc in enumerate(creators):
+            staff = Staff.objects.filter(work=work, role_id=nc["role_id"], artist=artists_list[index]).first()
```

Comment:
Could be improved even further.
You could do a `role_id__in=[nc["role_id"] for nc in creators]` query and do some set manipulations to exclude already created staff.

That'd bring down to: 1 query to get already existing stuff, 1 query to add everything.